internal/aliasgen/aliasgen.go

```diff
@@ -0,0 +1,421 @@
+	if err != nil {
+		return err
+	}
+	if err := am.WriteAliases(destDir); err != nil {
+		return err
+	}
+	if err := goImports(destDir); err != nil {
+		return err
+	}
+	if err := goModTidy(destDir); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Loads information about a Go package in the specified directory and returns
+// the information required to properly create aliases for the public surface.
+func createMappings(dir string) (*aliasGenerator, error) {
+	log.Printf("creating mappings for: %q", dir)
+	conf := &packages.Config{
+		Mode: packages.NeedName | packages.NeedTypes | packages.NeedDeps | packages.NeedSyntax,
+		Dir:  dir,
+	}
+
+	// Load all package info.
+	pkgs, err := packages.Load(conf)
+	if err != nil {
+		return nil, err
+	}
+	if len(pkgs) != 1 {
+		return nil, fmt.Errorf("found %d packages is %s, expected 1", len(pkgs), dir)
+	}
+	pkg := pkgs[0]
+	am := &aliasGenerator{
+		importPath: pkg.PkgPath,
+		pkg:        strings.TrimSuffix(pkg.Name, "pb"),
+	}
+
+	// Load corresponding documentation.
+	docPkg, err := doc.NewFromFiles(pkg.Fset, pkg.Syntax, pkg.PkgPath)
+	if err != nil {
+		return nil, err
+	}
+	identToDoc := make(map[string]string, len(docPkg.Types))
+	for _, t := range docPkg.Types {
+		identToDoc[t.Name] = t.Doc
+	}
+
+	// Copy information over for all public members.
+	for _, name := range pkg.Types.Scope().Names() {
+		obj := pkg.Types.Scope().Lookup(name)
+		if !obj.Exported() {
+			continue
+		}
+		switch obj.(type) {
+		case *types.Var:
+			am.vars = append(am.vars, obj.Name())
+		case *types.Const:
+			am.consts = append(am.consts, obj.Name())
+		case *types.TypeName:
+			am.typeNames = append(am.typeNames, &namedType{
+				name: obj.Name(),
+				doc:  identToDoc[obj.Name()],
+			})
+		case *types.Func:
+			f, err := processFunction(obj.(*types.Func))
+			if err != nil {
+				return nil, err
+			}
+			am.funcs = append(am.funcs, f)
+		default:
+			return nil, fmt.Errorf("unable to associate %q with type %T", obj.Name(), obj)
+		}
+	}
+	return am, nil
+}
+
+// processFunction parses types information from a function signature.
+func processFunction(f *types.Func) (*function, error) {
+	fn := &function{
+		name: f.Name(),
+	}
+	sig, ok := f.Type().(*types.Signature)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type %+v", f.Type())
+	}
+	for i := 0; i < sig.Params().Len(); i++ {
+		ti := &typeInfo{}
+		v := sig.Params().At(i)
+		ti.name = v.Name()
+		obj, isPtr, err := getTypeNameForFn(v.Type(), false)
+		if err != nil {
+			return nil, err
+		}
+		ti.typeName = obj.Name()
+		ti.pkg = obj.Pkg().Name()
+		ti.isPtr = isPtr
+		fn.params = append(fn.params, ti)
+	}
+	for i := 0; i < sig.Results().Len(); i++ {
+		ti := &typeInfo{}
+		v := sig.Results().At(i)
+		ti.name = v.Name()
+		obj, isPtr, err := getTypeNameForFn(v.Type(), false)
+		if err != nil {
+			return nil, err
+		}
+		ti.typeName = obj.Name()
+		ti.pkg = obj.Pkg().Name()
+		ti.isPtr = isPtr
+		fn.returns = append(fn.returns, ti)
+	}
```

Comment:
These could go in a function that takes a `*types.Tuple` and then called twice, once with `sig.Params()` and once with `sig.Results()` for the `fn.params` and `fn.returns` respectively. Thoughts?

Reply:
great idea.